src/executor/select.rs

```diff
@@ -997,31 +998,96 @@ where
     L: EntityTrait,
     R: EntityTrait,
 {
-    let mut acc: Vec<(L::Model, Vec<R::Model>)> = Vec::new();
-    for (l, r) in rows {
-        if let Some((last_l, last_r)) = acc.last_mut() {
-            let mut same_l = true;
-            for pk_col in <L::PrimaryKey as Iterable>::iter() {
-                let col = pk_col.into_column();
-                let val = l.get(col);
-                let last_val = last_l.get(col);
-                if !val.eq(&last_val) {
-                    same_l = false;
+    {
+        let pkcol = <L::PrimaryKey as Iterable>::iter()
+            .next()
+            .expect("should have primary key")
+            .into_column();
+
+        let keys: Vec<Value> = rows.iter().map(|row| row.0.get(pkcol)).collect();
+        let mut unique_keys: Vec<Value> = Vec::new();
+        for key in keys {
+            let mut existed = false;
+            for existing_key in unique_keys.clone() {
+                if key == existing_key {
+                    existed = true;
                     break;
                 }
             }
-            if same_l {
-                if let Some(r) = r {
-                    last_r.push(r);
-                    continue;
-                }
+            if !existed {
+                unique_keys.push(key)
             }
         }
-        let rows = match r {
-            Some(r) => vec![r],
-            None => vec![],
-        };
-        acc.push((l, rows));
+
+        let key_values: Vec<(Value, L::Model, R::Model)> = rows
+            .iter()
+            .map(|row| {
+                (
+                    row.0.clone().get(pkcol),
+                    row.0.clone(),
+                    row.1.clone().expect("should have linked entity"),
+                )
+            })
+            .collect();
+
+        let mut hashmap: HashMap<Value, (L::Model, Vec<R::Model>)> = key_values.into_iter().fold(
+            HashMap::<Value, (L::Model, Vec<R::Model>)>::new(),
+            |mut acc: HashMap<Value, (L::Model, Vec<R::Model>)>,
+             key_value: (Value, L::Model, R::Model)| {
+                acc.insert(key_value.0, (key_value.1, Vec::new()));
+
+                acc
+            },
+        );
+
+        rows.into_iter().for_each(|row| {
+            let key = row.0.get(pkcol);
+
+            let vec = &mut hashmap
+                .get_mut(&key)
+                .expect("Failed at finding key on hashmap")
+                .1;
+
+            vec.push(row.1.expect("should have value in row"));
+        });
+
+        let result: Vec<(L::Model, Vec<R::Model>)> = unique_keys
+            .into_iter()
+            .map(|key: Value| {
+                hashmap
+                    .get(&key)
+                    .cloned()
+                    .expect("should have key in hashmap")
+            })
+            .collect();
+        result
     }
-    acc
+
+    // let mut acc: Vec<(L::Model, Vec<R::Model>)> = Vec::new();
```

Comment:
if you want to keep the old code for reference, we can put it in another function `consolidate_query_result_of_ordered_rows`